### PR TITLE
Stop Serializing RepositoryData Twice when Writing

### PR DESCRIPTION
**Pull request by original-brownbear**
We can save one round of serializing `RepositoryData` on the write path.
This also leads to somewhat better compression because we compress larger chunks
in one go potentially when compared to serializing and compressing in one go.
Also, fixed the double wrapping of collections when copying the repository
data instance via the `withGenId`.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample-windows

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -1250,9 +1249,11 @@ private void doGetRepositoryData(ActionListener<RepositoryData> listener) {
                     loaded = repositoryDataFromCachedEntry(cached);
                 } else {
                     loaded = getRepositoryData(genToLoad);
-                    // We can cache in the most recent version here without regard to the actual repository metadata version since we're
-                    // only caching the information that we just wrote and thus won't accidentally cache any information that isn't safe
-                    cacheRepositoryData(loaded, Version.CURRENT);
+                    // We can cache serialized in the most recent version here without regard to the actual repository metadata version
+                    // since we're only caching the information that we just wrote and thus won't accidentally cache any information that
+                    // isn't safe
+                    cacheRepositoryData(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1255]: Still kind of awkward to right away serialize what we just read but I didn't want to make the bigger changes required for not having to serialize here in this PR + it's a rare spot to get here anyway and only applies to the first load of repository data on a master node.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -1601,7 +1601,7 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                             } catch (IOException e) {
                                 logger.warn(() -> new ParameterizedMessage("Failed to clean up old index blobs {}", oldIndexN), e);
                             }
-                            return writtenRepositoryData;
+                            return filteredRepositoryData.withGenId(newGen);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1604]: I think it's a bit weird to set this only at the very end. I think we shuld set the new genId on the RepositoryData object as soon as we have `final long newGen = setPendingStep.result();`?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Fair point, I pushed 602e32f413a to clean this up :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**